### PR TITLE
Add Docker Hub Images Tree View

### PR DESCRIPTION
**Pull request by jdneo**
1. add docker hub images tree view
2. can pull image from docker hub

### CODE REVIEW

File: src/dockerHubImages.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+                    this.context.asAbsolutePath(path.join("resources", "image.png")),
+                ));
+            });
+        } catch (error) {
+            if (!DockerTreeBase.isErrorMessageShown) {
+                vscode.window.showErrorMessage(`[Failed to list Docker Images] ${error.stderr}`);
+                DockerTreeBase.isErrorMessageShown = true;
+            }
+        } finally {
+            this.setAutoRefresh();
```
Comment (formulahendry) [inline src/dockerHubImages.ts]: As sync offline, we do not need setAutoRefresh.

File: package.json [hunk 2]
```diff
@@ -101,6 +105,18 @@
             {
                 "command": "docker-explorer.removeImage",
                 "title": "Remove"
+            },
+            {
```
Comment (formulahendry) [inline package.json]: Also add a 'refresh' button for user

File: src/DockerHub/DockerHubManager.ts [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+        .catch((error) => {
+            vscode.window.showErrorMessage(error);
+        });
+    }
+
+    public logout() {
+        this._userName = "";
+        this._password = "";
+        this._token = "";
+        this._images = [];
```
Comment (formulahendry) [inline src/DockerHub/DockerHubManager.ts]: Clean the image tree list after logout

File: src/DockerHub/DockerHubManager.ts [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+    private getUserCredential(): Promise<string[]> {
+        const userAndPassword = [];
+        return new Promise((resolve, reject) => {
+            vscode.window.showInputBox({prompt: "Enter user name."})
+            .then((val) => {
+                if (!val || val.trim().length === 0) {
+                    reject("Invalid user name.");
+                }
+                userAndPassword.push(val);
+                return vscode.window.showInputBox({prompt: "Enter password."});
```
Comment (formulahendry) [inline src/DockerHub/DockerHubManager.ts]: https://code.visualstudio.com/docs/extensionAPI/vscode-api#InputBoxOptions : {password: true}

File: package.json [hunk 5]
```diff
@@ -101,6 +105,18 @@
...
+                "command": "docker-explorer.loginDockerHub",
+                "title": "Login"
```
Comment (formulahendry) [inline package.json]: Add icon for 'login' and 'logout', you could do it later.

File: src/DockerHub/DockerHubManager.ts [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+    }
+
+    public get images(): string[] {
+        return this._images;
+    }
+
+    public login() {
+        this.getUserCredential()
+        .then((credential) => {
+            return this.sendLoginRequest(credential[0], credential[1]);
```
Comment (formulahendry) [inline src/DockerHub/DockerHubManager.ts]: Handle incorrect Credential

File: src/DockerHub/DockerHubManager.ts [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+                    reject("Error: Cannot get repositories.");
+                }
+            })
+            .catch((err) => {
+                reject(err);
+            });
+        });
+    }
+
+    private getTagsForRepos(repos: string[]): Promise<string[]> {
```
Comment (formulahendry) [inline src/DockerHub/DockerHubManager.ts]: As sync offline, only show repos in tree view list, and should 'tag' drop down list (vscode.window.showQuickPick) when pulling.

Comment (formulahendry) [inline src/DockerHub/DockerHubManager.ts]: BTW, the tag list would better still be cached

File: src/DockerHub/DockerHubManager.ts [hunk 8]
```diff
@@ -0,0 +1,162 @@
+import axios from "axios";
+import * as vscode from "vscode";
+import { Constants } from "./Constants";
+
+export class DockerHubManager {
+
+    private static _instance: DockerHubManager = new DockerHubManager();
+
+    private _userName: string = "";
```
Comment (formulahendry) [inline src/DockerHub/DockerHubManager.ts]: If user restart the VS Code, user need to login again. We may consider storing the credential in [WorkspaceConfiguration](https://code.visualstudio.com/docs/extensionAPI/vscode-api#WorkspaceConfiguration) or somewhere else.

Comment (LazarusX) [inline src/DockerHub/DockerHubManager.ts]: Our Arduino extension saves user configuration in ~/.vscode/arduino.json

File: src/dockerHubRepos.ts [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+        try {
+            for (const key of DockerHubManager.Instance.repos.keys()) {
+                repos.push(new DockerHubRepo(
+                    DockerHubManager.Instance.userName,
+                    key,
+                    this.context.asAbsolutePath(path.join("resources", "image.png")),
+                ));
+            }
+        } catch (error) {
+            if (!DockerTreeBase.isErrorMessageShown) {
```
Comment (formulahendry) [inline src/dockerHubRepos.ts]: If we do not enable auto-refresh, isErrorMessageShown is useless. We need always show error message. Just remove it.

File: src/DockerHub/DockerHubManager.ts (hunk 8 above)
Comment (formulahendry) [inline src/DockerHub/DockerHubManager.ts]: We need a 'global' configuration, while ~/.vscode/arduino.json is per workspace

File: src/extension.ts [hunk 10]
```diff
@@ -74,6 +79,22 @@ export function activate(context: vscode.ExtensionContext) {
         dockerImages.removeImage(image.repository, image.tag);
     }));
 
+    context.subscriptions.push(vscode.commands.registerCommand("docker-explorer.refreshDockerHubImages", () => {
```
Comment (formulahendry) [inline src/extension.ts]: Add BI for refreshDockerHubImages, loginDockerHub, logoutDockerHub

**Comment by formulahendry** [comment]
@jdneo After resolving the new two comments, you could 'Squash and merge' this PR. We could put 'icon' and 'credential' work in another PR.

File: src/DockerHub/DockerHubManager.ts (hunk 8 above)
Comment (LazarusX) [inline src/DockerHub/DockerHubManager.ts]: `~` means user directory, so `~/.vscode/arduino.json` == `C:\Users\alias\.vscode\arduino.json` or `/Users/alias/.vscode/arduino.json`, which is global.

Comment (formulahendry) [inline src/DockerHub/DockerHubManager.ts]: Got it. That's good. We could follow it as well.  cc @jdneo

**Pull request merged**